### PR TITLE
HTTP/3: Http3Stream pooling

### LINKED ISSUE #33680: HTTP/3: Cache and reusing streams
In HTTP/2 a connection responsible for creating and stream and all resources related to it. `Http2Stream` creation and reuse are all handled by `Http2Connection`.

In HTTP/3 a stream extends into the QUIC transport layer. `Http3Connection` creates `Http3Stream` and `QuicConnection` creates `QuicStreamContext`. For high performance, both these need to be cached and reused.

tldr; cache should move into the transport layer

Detail:

1. A concurrent queue is added to `QuicConnection`. It acts as a cache for completed `QuicStreamContext` instances.
2. When accepting a stream on a connection, the queue is checked for existing `QuicStreamContext` instances. An instance is reused if available. Otherwise, a new `QuicStreamContext` is created. The stream has a reference to its connection.
3. There isn't an API to "return" a stream to the connection on `MultiplexedConnectionContext`. Instead, returning will be implicit with disposing of `QuicStreamContext`. When a cleanly completed (i.e. neither side of the stream was aborted) stream is disposed it will add itself to its `QuicConnection` concurrent queue cache so it can be reused.
4. We're now reusing the `QuicStreamContext` 🥳 We also want to reuse the `Http3Stream`. Rather than having a separate cache, the feature collection on the stream transport - `QuicStreamContext.FeatureCollection` - is used to stash `Http3Stream`. A reused `QuicStreamContext` will have a `Http3Stream` on it to reuse.

Caching and reusing streams should only be necessary for request streams (i.e. bidirectional streams) on an HTTP/3 connection. There isn't value in caching and reusing unidirectional streams. Only a handful are created for a connection and they will live for the lifetime of the connection.

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/33680

This builds on top of the new `IPersistentStateFeature`.

Http3Stream and many of its associated types (pipes, output producer, decoder) are now reused between requests.

In this PR I also added some basic microbenchmarks that do HTTP/3. Refactoring the test code so it is usable in microbenchmarks is the major of line changes in this PR.

I will get benchmarks comparing allocations before and after.

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]
Seems OK in general once the build is fixed. I'm curious to see the benchmarks.

File: src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs [hunk 1]
```diff
@@ -257,36 +259,37 @@ private void UpdateStartingStreams(DateTimeOffset now)
...
-                            var stream = new Http3ControlStream<TContext>(application, httpConnectionContext);
+                            var stream = new Http3ControlStream<TContext>(application, CreateHttpStreamContext(streamContext));
                             _streamLifetimeHandler.OnStreamCreated(stream);
 
                             ThreadPool.UnsafeQueueUserWorkItem(stream, preferLocal: false);
                         }
                         else
                         {
+                            var persistentStateFeature = streamContext.Features.Get<IPersistentStateFeature>();
+                            Debug.Assert(persistentStateFeature != null, $"Required {nameof(IPersistentStateFeature)} not on stream context.");
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs:273]: This seems too possible to just assert given the transport is pluggable. Let's null-check and throw.

**Review by halter73 — APPROVED** [review]
I also look forward to the benchmarks.

**Review by dougbu — CHANGES_REQUESTED** [review]
New and renamed files need MIT license headers

**Comment by JamesNK** [comment]
I thought I had already updated old files. Which ones still need to be updated?

**Review by dougbu — APPROVED** [review]
> I thought I had already updated old files. Which ones still need to be updated?

I must have looked at this and still had an older copy of the changes cached. Sorry.

I don't see another way to remove my changes request and am signing off only on the headings.

**Comment by JamesNK** [comment]
Before - 300mb allocated for 10,000 requests
image

After - 50mb allocated for 10,000 requests
image

**Pull request merged**